Sources/Tracking/Service/HttpClient.swift

```diff
@@ -165,4 +174,74 @@ extension CIOHttpClient {
         #endif
         return userAgent
     }
+
+    // In certain scenarios, it makes sense for us to pause making any HTTP requests to the
+    // Customer.io API. Because HTTP requests are performed by the background queue, there is
+    // a chance that the background queue could make a lot or more HTTP requests in
+    // a short amount of time from a device which makes a performance impact on our API.
+    // By pausing HTTP requests, we mitigate the chance of customer devices causing harm to our API.
+    private func pauseHttpRequests() {
+        // We do *not* want to make this value configurable by the customer because
+        // we are pausing the SDK to protect our own API from spamming results.
+        let minutesToPause = 5
+        let dateToEndPause = Date().addMinutes(minutesToPause)
+
+        globalDataStore.httpRequestsPauseEnds = dateToEndPause
+
+        logger.info("All HTTP requests to the Customer.io API have been paused for \(minutesToPause) minutes.")
+    }
+
+    /**
+     - When receiving a 5xx response:
+     * Begin an exponential backoff retry on the HTTP task that returned back the 5xx error.
+     * After these retry attempts, if the HTTP request is still receiving a 5xx response then the
+       requests will sleep for 5 minutes and no requests will be attempted.
+     * After the 5 minutes, HTTP requests are able to be run as normal. No memory of any errors prior.
+
+     - When receiving a 401 response:
+     * The HTTP requests will sleep for 5 minutes as above.
+
+     - Any other 4xx error
+     * Log the error as it's more then likely a SDK developer error or an error by the customer.
+     */
+    private func handleUnsuccessfulStatusCodeResponse(
+        statusCode: Int,
+        data: Data?,
+        params: HttpRequestParams,
+        onComplete: @escaping (Result<Data, HttpRequestError>) -> Void
+    ) {
+        let unsuccessfulStatusCodeError: HttpRequestError =
+            .unsuccessfulStatusCode(statusCode,
+                                    apiMessage: getErrorMessage(responseBody: data))
+
+        switch statusCode {
+        case 500 ..< 600:
+            if let sleepTimeMilliseconds = retryPolicy.nextSleepTimeMilliseconds {
+                logger
+                    .debug("""
+                    Encountered \(statusCode) HTTP response.
+                    Sleeping \(sleepTimeMilliseconds) and then retrying.
+                    """)
+
+                retryPolicyTimer.schedule(milliseconds: sleepTimeMilliseconds) {
+                    self.request(params, onComplete: onComplete)
```

Comment:
creates a whole different pathway to a request being called, `QueueRunRequest` might be a better context to be handling this sort of logic, versus having a completely separate scheduler here. My imagining was:

- error bubbles back up to the queue. 500 means server unhappy, stop the queue for now
- retries are handled by rescheduling the entire bg queue to the next exponential backoff time, following the retry policy, which then is reset when we get a non-500 / get paused

In this scenario, for 100 unrelated requests in the queue, would we end up spawning 100 timers on their own retry loops? or is this a blocking call?

Reply:
https://github.com/customerio/customerio-ios/blob/levi/background-q-errorhandling/Sources/Tracking/Background%20Queue/QueueRunRequest.swift#L80

this was mostly what I got concerned about. From reading just the HTTPClient code, it's not clear that the queue won't continue until `onComplete` is called, which made me concerned that the bg queue might move ahead while this gets deferred and get us into a bit of a mess.

I still don't love having the retry policy & associated timer hidden inside the HttpClient, versus handling this higher up in the call stack, but my concern about multiple spawned timers is not a problem at least.

Reply:
fine to leave as-is for now, can revisit the architecture of it at a later date if need be